### PR TITLE
added everything, results don't reproduce

### PR DESCRIPTION
**Pull request by ricomnl**

### CODE REVIEW

File: scripts/add_a_method.sh
Comment (szalata) [inline scripts/add_a_method.sh:1]: I suppose the modification of this file is accidental

File: .gitmodules [hunk 1]
```diff
@@ -0,0 +1,3 @@
+[submodule "src/task/methods/third_place/single_cell_perturbations"]
+	path = src/task/methods/third_place/single_cell_perturbations
+	url = https://github.com/okon2000/single_cell_perturbations/
```
Comment (rcannood) [inline .gitmodules:3]: Please remove this submodule. Possible solutions are:

* If the resources you need are <10KB, copy the scripts you need to `src/task/methods/third_place/single_cell_perturbations/resources` and add them to `functionality.resources`
* Install the dependencies you need inside the Viash component itself:
  ```yaml
  platforms:
  - type: docker
    image: ...
    setup:
    - type: python
      packages: git+...
  ```
* Clone the repository inside your docker image
  ```yaml
  platforms:
  - type: docker
    image: ...
    setup:
    - type: apt
      packages: git
    - type: docker
      run: git clone https://github.com/okon2000/single_cell_perturbations/ /var/single_cell_perturbations
  ```

File: scripts/add_a_method.sh [hunk 2]
```diff
@@ -8,7 +8,7 @@ exit 1
 scripts/download_resources.sh
 
 # create a new component
-method_id="my_method"
+method_id="third_place"
```
Comment (rcannood) [inline scripts/add_a_method.sh:11]: Undo this change

```suggestion
method_id="my_method"
```

File: src/task/methods/third_place/config.vsh.yaml [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+  info:
+    # A relatively short label, used when rendering visualisations (required)
+    label: Third Place
+    # A one sentence summary of how this method works (required). Used when 
+    # rendering summary tables.
+    summary: "FILL IN: A one sentence summary of this method."
+    # A multi-line description of how this component works (required). Used
+    # when rendering reference documentation.
+    description: |
+      FILL IN: A (multi-line) description of how this method works.
```
Comment (rcannood) [inline src/task/methods/third_place/config.vsh.yaml:23]: Please fill in this information.

File: src/task/methods/third_place/config.vsh.yaml [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    # Which normalisation method this component prefers to use (required).
+    preferred_normalization: log_cp10k
+    # A reference key from the bibtex library at src/common/library.bib (required).
+    reference: bibtex_reference_key
+    # URL to the documentation for this method (required).
+    documentation_url: https://url.to/the/documentation
+    # URL to the code repository for this method (required).
+    repository_url: https://github.com/organisation/repository
```
Comment (rcannood) [inline src/task/methods/third_place/config.vsh.yaml:31]: Please also fill in this information

File: src/task/methods/third_place/config.vsh.yaml [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+  # Resources required to run the component
+  resources:
+    # The script of your component (required)
+    - type: python_script
+      path: script.py
+    - path: single_cell_perturbations/util.py
+    - path: single_cell_perturbations/models.py
+    - path: single_cell_perturbations/config/params_stage_1.py
+    - path: single_cell_perturbations/config/params_stage_2.py
+    - path: single_cell_perturbations/models/simple/model_7/model_7_0.keras
```
Comment (rcannood) [inline src/task/methods/third_place/config.vsh.yaml:49]: How was this model created? Please also create a component for generating this model. If you are pretraining models, `src/task/methods/third_place/config.vsh.yaml` should inherit from `comp_method_run.yaml` and not `comp_method.yaml`, and `src/task/methods/third_place_train/config.vsh.yaml` should inherit from `comp_method_train.yaml`.

**Review by rcannood — CHANGES_REQUESTED** [review]
Hey @ricomnl ! I like that the code in your PR looks very clean already :ok_hand: .

It seems this method uses a pre-trained model. As such; I'd like to ask to create a second component which trains the model.

Also, I'd prefer if you didn't use a submodule but instead install or clone the repository inside the Docker setup, or copy the files you need from the original repository.

Let me know if you have any questions or when you want me to rereview

**Comment by ricomnl** [comment]
> It seems this method uses a pre-trained model. As such; I'd like to ask to create a second component which trains the model.

@rcannood we used the pre-trained model as a test, I'll remove it from the config. The code itself trains a new model but it currnetly doesn't reproduce the authors' original results

**Comment by rcannood** [comment]
> we used the pre-trained model as a test, I'll remove it from the config. 

Ah ok, great!

> The code itself trains a new model but it currnetly doesn't reproduce the authors' original results

I suppose we should contact the authors, then?

**Comment by szalata** [comment]
@ricomnl Could you please compare the output with the submission file generated by this reproducibility notebook by the authors? https://www.kaggle.com/code/jankowalski2000/3rd-place-solution/output
The notebook should correspond to the method from the repository as well

File: .gitmodules (hunk 1 above)
Comment (ricomnl) [reply .gitmodules:3]: done

File: scripts/add_a_method.sh (hunk 2 above)
Comment (ricomnl) [reply scripts/add_a_method.sh:11]: got it

**Comment by ricomnl** [comment]
@rcannood it's almost reproducing the authors submission now when run via `python script.py` but there are some issues that come up when running it with viash that I haven't had time to fully debug yet

File: src/task/methods/third_place/script.py [hunk 6]
```diff
@@ -0,0 +1,750 @@
...
+from tensorflow.keras.optimizers.legacy import Adam
+from sklearn.decomposition import TruncatedSVD
+
+## VIASH START
+par = {
+    "de_train": "resources/neurips-2023-kaggle/de_train.parquet",
+    "de_test": "resources/neurips-2023-kaggle/de_test.parquet",
+    "id_map": "resources/neurips-2023-kaggle/id_map.csv",
+    "output": "output.parquet",
+}
```
Comment (rcannood) [inline src/task/methods/third_place/script.py:24]:
```suggestion
par = {
    "de_train": "resources/neurips-2023-kaggle/de_train.parquet",
    "id_map": "resources/neurips-2023-kaggle/id_map.csv",
    "output": "output.parquet",
}
meta = {
    "resources_dir": "src/task/methods/third_place"
}
```

File: src/task/methods/third_place/script.py [hunk 7]
```diff
@@ -0,0 +1,750 @@
...
+        "sm_name",
+        "sm_lincs_id",
+        "SMILES",
+        "split",
+        "control",
+        "index",
+    }
+]
+
+pseudolabel = pd.read_csv("submission(14).csv")
```
Comment (rcannood) [inline src/task/methods/third_place/script.py:711]:
```suggestion
pseudolabel = pd.read_csv(meta["resources_dir"] + "/submission(14).csv")
```

File: src/task/methods/third_place/script.py [hunk 8]
```diff
@@ -0,0 +1,750 @@
...
+new_names = le.transform(original_x.flat).reshape(-1, 2)
+
+x_test = le.transform(test_df[["cell_type", "sm_name"]].values.flat).reshape(-1, 2)
+
+preds = []
+for model, param in zip(models, params):
+    temp_pred = [
+        fit_and_predict_embedding_nn(new_names, original_y, x_test, model, param)
+        for i in range(reps)
+    ]
```
Comment (rcannood) [inline src/task/methods/third_place/script.py:733]: This code throws an error for me:

    Reading input files
    Traceback (most recent call last):
      File "/tmp/viash-run-third_place-hMmNEs.py", line 748, in <module>
        temp_pred = [
                    ^
      File "/tmp/viash-run-third_place-hMmNEs.py", line 749, in <listcomp>
        fit_and_predict_embedding_nn(new_names, original_y, x_test, model, param)
      File "/tmp/viash-run-third_place-hMmNEs.py", line 696, in fit_and_predict_embedding_nn
        y = d.fit_transform(y)
            ^^^^^^^^^^^^^^^^^^
      File "/usr/local/lib/python3.11/site-packages/sklearn/utils/_set_output.py", line 157, in wrapped
        data_to_wrap = f(self, X, *args, **kwargs)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/usr/local/lib/python3.11/site-packages/sklearn/base.py", line 1152, in wrapper
        return fit_method(estimator, *args, **kwargs)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/usr/local/lib/python3.11/site-packages/sklearn/decomposition/_truncated_svd.py", line 229, in fit_transform
        X = self._validate_data(X, accept_sparse=["csr", "csc"], ensure_min_features=2)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/usr/local/lib/python3.11/site-packages/sklearn/base.py", line 605, in _validate_data
        out = check_array(X, input_name="X", **check_params)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/usr/local/lib/python3.11/site-packages/sklearn/utils/validation.py", line 957, in check_array
        _assert_all_finite(
      File "/usr/local/lib/python3.11/site-packages/sklearn/utils/validation.py", line 122, in _assert_all_finite
        _assert_all_finite_element_wise(
      File "/usr/local/lib/python3.11/site-packages/sklearn/utils/validation.py", line 171, in _assert_all_finite_element_wise
        raise ValueError(msg_err)
    ValueError: Input X contains NaN.

Do you get the same error?

Comment (ricomnl) [reply src/task/methods/third_place/script.py:733]: yes, it's super odd. If I run the file as script outside of viash I don't get the error

Comment (rcannood) [reply src/task/methods/third_place/script.py:735]: It took me too long to figue out :)

When you run the component with `resources/neurips-2023-kaggle/de_train.parquet`, everything works -- which is what you're doing.

However, when running `viash test`, the component is also run with `resources/neurips-2023-data/de_train.parquet`, which has more genes than the `neurips-2023-kaggle` does.

Since the current implementation relies on `path: submission(14).csv`, this causes errors. 

According to the discussion at https://www.kaggle.com/competitions/open-problems-single-cell-perturbations/discussion/458750, `submission(14).csv` is created by version 264 of the notebook they provide. Do you think so, too?

**Review by rcannood — APPROVED** [review]
LGTM!
